### PR TITLE
feat(*): hijack application metrics

### LINKED ISSUE #3457: Proxy application metrics through envoy
### Description

Currently the application metrics are exposed as a regular service.
This is tricky if you want prometheus metrics scrape to not use mTLS.
See doc: https://kuma.io/docs/1.4.0/policies/traffic-metrics/#expose-metrics-from-applications

What could be done is for envoy to proxy metrics from the application as a service similar to the prometheus one.
This way we could easily use [kuma-sd](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kuma_sd_config) to also expose application metrics and not risk customers to accidentally expose more than necessary.

Once this is done it will probably make sense to disable mTLS on the listener exposing metrics by default (most of the time prometheus scrape is in plaintext).

It's currently unclear how this should be setup and requires some designing.

### PR DESCRIPTION
**Pull request by lukidzi**
### Summary

Currently, to expose applications metrics Prometheus needs to communicate using mtls to scrape metrics from the application. This PR introduce change that allows to get application metrics without mTLS. 

### Full changelog

* [Bootstrap request now return different model but CP support old one]
* [Implement concurrent scraping metrics in kuma-dp]
* [Implement generate and send configuration]

### Issues resolved

Fix #3457

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### Testing

- [X] Unit tests
- [X] E2E tests
- [X] Manual testing on Universal
- [X] Manual testing on Kubernetes

### Backwards compatibility

- [ ] Update [`UPGRADE.md`](/UPGRADE.md) with any steps users will need to take when upgrading.
- [ ] Add `backport-to-stable` label if the code follows our [backporting policy](/CONTRIBUTING.md#backporting)

### CODE REVIEW

File: app/kuma-dp/cmd/testdata/bootstrap-config.golden.yaml [hunk 1]
```diff
@@ -17,6 +17,11 @@ dynamicResources:
 node:
   cluster: backend
   id: dp-1.default.default
+  metadata:
+    dataplane.applications.metrics:
+    - name: app
+      port: 123
+      path: "/stats"
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/cmd/testdata/bootstrap-config.golden.yaml:24]: we don't need this in metadata anymore, right?

File: api/mesh/v1alpha1/metrics.proto [hunk 2]
```diff
@@ -48,4 +48,22 @@ message PrometheusMetricsBackendConfig {
...
   // is enabled in Mesh. If nil, then it is treated as false.
   google.protobuf.BoolValue skipMTLS = 4;
+
+  // Map with the configuration of applications which metrics are going to be
+  // scrapped by kuma-dp.
+  map<string, PrometheusAggregateMetricsConfig> aggregate = 5;
+}
+
+// PrometheusAggregateMetricsConfig defines endpoints that should be scrapped
+// by kuma-dp for prometheus metrics.
```
Comment (jakubdyszkiewicz) [inline api/mesh/v1alpha1/metrics.proto:58]: I'd leave a comment that changing this setting requires a sidecar restart.

File: pkg/xds/bootstrap/handler.go [hunk 3]
```diff
@@ -49,22 +49,36 @@ func (b *BootstrapHandler) Handle(resp http.ResponseWriter, req *http.Request) {
...
+	if req.Header.Get("accept") == "application/json" {
+		resp.Header().Set("content-type", "application/json")
+		response := createBootstrapResponse(bootstrapBytes, params)
+		responseBytes, err = json.Marshal(response)
+		if err != nil {
+			logger.Error(err, "Could not convert to json")
+			resp.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	} else {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/handler.go:75]: add a comment `// backwards compatibility`

File: test/e2e_env/kubernetes/metrics/applications_metrics.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+		pods, err := k8s.ListPodsE(
+			env.Cluster.GetTesting(),
+			env.Cluster.GetKubectlOptions(namespace),
+			metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", "test-server"),
+			},
+		)
+		Expect(err).ToNot(HaveOccurred())
+		clientPodName := pods[0].Name
+		podIp := pods[0].Status.PodIP
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/kubernetes/metrics/applications_metrics.go:124]: Create `PodIPForApp` just like `PodNameOfApp`

File: pkg/xds/bootstrap/generator.go [hunk 5]
```diff
@@ -27,7 +27,7 @@ import (
 )
 
 type BootstrapGenerator interface {
-	Generate(ctx context.Context, request types.BootstrapRequest) (proto.Message, error)
+	Generate(ctx context.Context, request types.BootstrapRequest) (proto.Message, *configParameters, error)
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator.go:30]: we need a better type here. configParameters is private, the interface should only return public types.

File: app/kuma-dp/cmd/testdata/bootstrap-config.golden.yaml (hunk 1 above)
Comment (lukidzi) [reply app/kuma-dp/cmd/testdata/bootstrap-config.golden.yaml]: Removed

File: pkg/xds/bootstrap/testdata/generator.metrics-aggregate-config.kubernetes.golden [hunk 6]
```diff
@@ -0,0 +1,115 @@
+dynamicResources:
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/testdata/generator.metrics-aggregate-config.kubernetes.golden:1]: `.yaml` is missing in the filename

**Review by jakubdyszkiewicz — COMMENTED** [review]
bunch of nits, but overall 👍

File: app/kuma-dp/pkg/dataplane/metrics/server.go [hunk 7]
```diff
@@ -85,34 +98,69 @@ func (s *Hijacker) Start(stop <-chan struct{}) error {
...
+		if _, err := writer.Write(resp); err != nil {
+			logger.Error(err, "error while writing the response")
+		}
+		if _, err := writer.Write([]byte("\n")); err != nil {
+			logger.Error(err, "error while writing the response")
+		}
+	}
+	// if _, err := writer.Write(buf.Bytes()); err != nil {
+	// 	logger.Error(err, "error while writing the response")
+	// }
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/metrics/server.go:137]: remove

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**